From 98d74e8b47852acb00ba3557eafc9a5801812e04 Mon Sep 17 00:00:00 2001
From: Aaron Plattner <aplattner@nvidia.com>
Date: Wed, 27 Oct 2021 21:45:43 -0700
Subject: [PATCH 4/6] CI: Switch from archlinux/base:latest to archlinux:latest

Signed-off-by: Aaron Plattner <aplattner@nvidia.com>
---
 .gitlab-ci.yml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/.gitlab-ci.yml b/.gitlab-ci.yml
index 11fb81a..285774f 100644
--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -1,4 +1,4 @@
-image: archlinux/base:latest
+image: archlinux:latest
 
 stages:
  - build
-- 
2.34.1

